src/reader_disc.cpp

```diff
@@ -0,0 +1,559 @@
+    for (i = 0; i < 16; i++) {
+      if (psz_serial[i] < '0' || psz_serial[i] > '9') {
+        char psz_tmp[16 + 1];
+        sprintf(psz_tmp, "%.2x%.2x%.2x%.2x%.2x%.2x%.2x%.2x", psz_serial[0],
+                psz_serial[1], psz_serial[2], psz_serial[3], psz_serial[4],
+                psz_serial[5], psz_serial[6], psz_serial[7]);
+        memcpy(psz_serial, psz_tmp, 16);
+        break;
+      }
+    }
+
+    /* Get disk key, since some discs have got same title, manufacturing
+     * date and serial number, but different keys */
+    if (dev->media.dvdcss.protection == 0x01) {
+      psz_key[0] = '-';
+      for (i = 0; i < DVD_KEY_SIZE; i++) {
+        sprintf(&psz_key[1 + i * 2], "%.2x", dev->media.dvdcss.DK[i]);
+      }
+      psz_key[1 + DVD_KEY_SIZE * 2] = '\0';
+    } else {
+      psz_key[0] = 0;
+    }
+
+    /* We have a disc name or ID, we can create the cache dir */
+    i = sprintf(dev->media.dvdcss.psz_cachefile, "%s", psz_cache);
+    if (mkdir(dev->media.dvdcss.psz_cachefile, 0755) < 0 && errno != EEXIST) {
+      printf("CSS: failed creating cache directory\n");
+      dev->media.dvdcss.psz_cachefile[0] = '\0';
+      goto nocache;
+    }
+
+    i += sprintf(dev->media.dvdcss.psz_cachefile + i, "/%s-%s%s", psz_title,
+                 psz_serial, psz_key);
+    if (mkdir(dev->media.dvdcss.psz_cachefile, 0755) < 0 && errno != EEXIST) {
+      printf("CSS: failed creating cache subdirectory\n");
+      dev->media.dvdcss.psz_cachefile[0] = '\0';
+      goto nocache;
+    }
+    i += sprintf(dev->media.dvdcss.psz_cachefile + i, "/");
+
+    /* Pointer to the filename we will use. */
+    dev->media.dvdcss.psz_block = dev->media.dvdcss.psz_cachefile + i;
+
+    printf("CSS: using CSS key cache dir: %s\n",
+           dev->media.dvdcss.psz_cachefile);
+  }
+nocache:
+  return 0;
+}
+#endif
+
+/* Loop over all titles and call dvdcss_title to crack the keys. */
+static int initAllCSSKeys(drive_info *dev) {
+  struct timeval all_s, all_e;
+  struct timeval t_s, t_e;
+  char filename[MAX_UDF_FILE_NAME_LEN];
+  unsigned int start, len;
+  int title;
+  //  int stitle;
+  //  bool st0;
+
+  udf_t udf;
+  /*
+    char *nokeys_str = getenv("DVDREAD_NOKEYS");
+    if(nokeys_str != NULL)
+      return 0;
+  */
+  udf.dev = dev;
+  udf.cache = NULL;
+
+  if (!dev->silent) {
+    printf("libdvdread: Attempting to retrieve all CSS keys\n");
+  }
+
+  gettimeofday(&all_s, NULL);
+
+  for (title = 0; title < 100; title++) {
+    //	stitle=0; st0=0;
+    gettimeofday(&t_s, NULL);
+    if (title == 0) {
+      sprintf(filename, "/VIDEO_TS/VIDEO_TS.VOB");
+    } else {
+      sprintf(filename, "/VIDEO_TS/VTS_%02d_%d.VOB", title, 0);
+    }
+    start = UDFFindFile(&udf, filename, &len);
+    if (start != 0 && len != 0) {
+      /* Perform CSS key cracking for this title. */
+      //	  st0=1;
+      if (!dev->silent) {
+        printf("libdvdread: Get key for %s at 0x%08x, len %10d\n", filename,
```

Comment:
Codacy found an issue: [%d in format string (no. 3) requires 'int' but the argument type is 'unsigned int'.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)